### ISSUE TITLE
BUG: Bilinear dtype promotion

### ISSUE THREAD

**Issue by cpelley** [SciTools/iris#2058]
Currently bilinear regridding gives unexpected results for source data which is of boolean or unsigned integer type due to the way that type promotion is currently handled [here](https://github.com/SciTools/iris/blob/de2cac5534a746053148d776afc4c564f8c458f4/lib/iris/analysis/_regrid.py#L223)

An easy fix is the following:

``` python
dtype = np.promote_types(dtype, 'float16')
```

**Comment by cpelley** [comment]
Any reasons against this change before I make the change?

**Comment by bjlittle** [comment]
@cpelley Are you suggesting that you promote the dtype regardless?

**Comment by cpelley** [comment]
Regardless of what sorry?

I should clarify what this does and why this makes sense:

```python
>>> np.promote_types('int64', 'float16')
dtype('float64')
>>> np.promote_types('int32', 'float16')
dtype('float64')
>>> np.promote_types('int8', 'float16')
dtype('float16')
>>> np.promote_types('uint8', 'float16')
dtype('float16')
>>> np.promote_types('bool', 'float16')
dtype('float16')
```

So this change means the resulting array that bilinear populates is always a float array (taking the minimum kind allowed).
Currently bilinear regrid can return a boolean array when the source is boolean.  It makes no sense for bilnear regrid to return a boolean array.  This extends the current existing thought that integer type is cast as 16-bit float

**Comment by bjlittle** [comment]
@cpelley What I meant was, are you simply proposing to replace
```python
if method = 'linear':
    if dtype.kind == 'i':
        dtype = np.promote_types(dtype, np.float16)
```

with
```python
if method == 'linear':
    dtype = np.promote_types(dtype, np.float16)
```

i.e. execute the `promote_types` regardlessly.

**Comment by bjlittle** [comment]
The change sound sensible to me, it'll be interesting whether it breaks any tests though ...

**Comment by cpelley** [comment]
Ah OK, yes that is what I meant :)

Thanks for your input @bjlittle, I'll get onto it!

**Comment by pelson** [comment]
@cpelley - is this still an issue. I've just closed the experiment PR in #2088, but there has been a merge from #2203. Does that address the bug being raised here? If not, would you like to revive #2088 and we will get it in? Otherwise, I don't think this issue needs to hang around.

Cheers,
